scicloj/tablecloth
# Add initial column API functions

## Goal

This PR aims just to add the start of a basic API related to enhancing support for the `column`. So we want to be able have some basic functions for creating columns, for knowing that they are columns, and what's in them. 

## Solution Overview

The PR adds a new namespace/api: `tablecloth.column.api`. As @genmeblog said, this separate API can help provide a "separation between dataset operations vs column operations." 

Within this new namespace, we start with these functions:
* `column` - creates a column from a vector or list
* `column?` - determines whether an item is a column or not
* `typeof` - returns the data type of the column's elements
*  `typeof?` - check to see if elements match a datatype
*  `zeros` - returns a column filled with zeros
*  `ones` - returns a column filled with ones

The ideas behind this small set of functions are as follows:
* We are basically creating a new API for `columns` whereas up until now `tablecloth` has been only about the dataset.
* We want to be able to create a column independently of a `dataset`, hence `column`. This gives the `column` some independence.
* Since we have an independent `column`, we should be able to identify that an item is a column, hence `column?`.
* Columns are uniformly typed entities (unless they are mixed in which case they are uniformly filled with `:object`). So we should be able to identify the datatype of the column's elements: `typeof`. This naming mimics R's vectors.
* From numpy & matlab, we take the function names `zeros` & `ones` for functions that can fill a column with some values. 

## How to test

There's a test suite add with this PR @ `test/tablecloth/column/api/column_test.clj`.
1. Take look to see if they tests are making sense
2. Run tests: `lein midje tablecloth.column.api.column-test`
3. - [ ] confirm tests pass
